### ISSUE TITLE
TabController scroll exception when deleting tabs in a dynamic tab layout

### ISSUE THREAD

**Issue by ZhennanWu** [flutter/flutter#24424]
When there are two tabs left and you delete the second tab, TabController throws an exception complaining it can not perform scroll animation when length<2, which is reasonable (I'm not complaining about this). However this exception cannot be avoided with regular methods as far as I know. 

 If you delete the first one, since there is no scroll animation, it's perfectly fine. TabController.animateTo(index-1) before deletion could have solved this exception, but it didn't. Since the animation is apparently async and is performed after the code complete.

It causes some burden when developing app with dynamic tabs. Right now TabController seems not quite fit into this scenario. Or is it I'm using it the wrong way and there's better way to implement this? Please let me know.

## Steps to Reproduce

1. Run the demo code below
2. Create a few tabs, like "0", "1", "2", "3"
3. Delete tabs **from back to front**. Delete "3", "2", "1"
4. When deleting "1", the exception is thrown
5. tabController.animateTo(pos-1) has no effect, until a delay is introduced (see in the demo code).

## Logs

```
I/flutter ( 5137): The following assertion was thrown while notifying listeners for AnimationController:
I/flutter ( 5137): 'package:flutter/src/material/tab_controller.dart': Failed assertion: line 175 pos 12: 'length > 1':
I/flutter ( 5137): is not true.
...
I/flutter ( 5137): When the exception was thrown, this was the stack:
I/flutter ( 5137): #2      TabController.offset= (package:flutter/src/material/tab_controller.dart:175:12)
I/flutter ( 5137): #3      _TabBarViewState._handleScrollNotification (package:flutter/src/material/tabs.dart:1173:19)
```

## Demo code (main.dart) 

```dart
import 'package:flutter/material.dart';
import 'dart:async';
void main() => runApp(MyApp());
class MyApp extends StatefulWidget {
  @override MyAppState createState() => MyAppState();
}
class MyAppState extends State<MyApp> with TickerProviderStateMixin {
  List<Tab> tabs = [];TabController tabController;var count = 1;
  void newTab() {
    setState(() {
      tabs.add(Tab(text: '$count',));count++;
      tabController = TabController(length: tabs.length, vsync: this);
    });
  }
  void closeCurrentTab() {
    final pos=tabController.index;
    setState((){
      if(pos>0)
        tabController.animateTo(pos-1);//This COULD have solved the problem.
    });
    //However, a delay must be introduced in order to let animateTo take effect. Uncomment to let delay take effect.
    //Timer(kTabScrollDuration,(){
      setState(() {
        tabs.removeAt(pos);
        tabController = TabController(initialIndex: pos-1,length: tabs.length, vsync: this);
      });
    //});

  }
  @override void initState() {
    super.initState();
    tabs.add(Tab(text: '0',));
    tabController = TabController(length: tabs.length, vsync: this);
  }
  @override void dispose() {
    tabController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          actions: <Widget>[
            IconButton(icon: Icon(Icons.add), onPressed: newTab),
            IconButton(icon: Icon(Icons.close), onPressed: closeCurrentTab,)
          ],
          bottom: TabBar(controller: tabController, tabs: tabs.map((tab) => tab).toList()),
        ),
        body: TabBarView(controller: tabController, children: tabs.map((tab) => Text(tab.text)).toList()),
      ),
    );
  }
}

```

**Comment by zoechi** [comment]
Please add the output of `flutter doctor -v`.

**Comment by ZhennanWu** [comment]
Good, seems I'm not the only one. #18481 is identical to this issue.

**Comment by HansMuller** [comment]
This bug is not a duplicate of #18481, which has been fixed.

**Comment by HansMuller** [comment]
This bug can be worked around for now by only resetting the tab controller after it has stopped changing. Here's a version of the original test case that does that: https://gist.github.com/HansMuller/15d7f306139ec92e5ecefb58f85a3d3f

**Comment by HansMuller** [comment]
This is a simplified version of the test case. It still possible to make it trip the length assert (apparently due to a pending TabBarView scroll being handled after the tab controller has been replaced) but, FWIW, it's much more difficult to make it fail. 

https://gist.github.com/HansMuller/37abc036d293bfc3ca27b943127a5689

**Comment by HansMuller** [comment]
Fixed in #24892
